### PR TITLE
Split HashJoin to SimpleHashJoin and CompoundHashJoin

### LINKED ISSUE #8342: Potential optimization for in-memory Join of a big table with a small table
Currently, our `HashJoin` strategy builds a `HashMap` for both the left and right table that are being joined.

This is needed to efficiently handle mixed join conditions where `Equals` conditions are mixed with `Between` conditions.

However, in case where we only have `Equals` conditions and no `Between` conditions, it may sometime be beneficial to only create the 'index' hashmap for one of the two tables (preferably the smaller one of the two, to ensure it is cheaper to build) - then we can just linearly iterate through the rows of the bigger table and correlate rows based on the smaller table's index.

This has the same asymptotic cost, but it should allow us to cut the cost of building the index for the bigger table - so it could be a bit faster in practice (very much depending on scenarios).

> This was originally planned as part of #8217 but it has been split off as a separate task, because it has slightly lower priority. It may still be an interesting experiment to try.

- [ ] Add a benchmark to the `Join.enso` suite of benchmarks that tests joining a relatively small table with a very big table (e.g. 100 to 10^6).
	- The benchmarks should probably be run only if `extended_tests` is enabled.
	- Ideally, we may want to check both `small.join big` and `big.join small` to check if the improvement does not 'care' about order in which the tables are given.
- [ ] Implement a `SimpleHashJoin` strategy that can be used if there are no `Between` conditions.
	- We can rename current `HashJoin` to `CompoundHashJoin` (because this is the one that allows mixing it with nested strategies).
	- Within `SimpleHashJoin`, compute the index for only the smaller of the two tables. Adapt the logic for finding matched and unmatched rows accordingly.
- [ ] Compare the benchmarks before and after the change.

### PR DESCRIPTION
**Pull request by AdRiley**
### Pull Request Description

Completes #8342 . Creates a SimpleHashJoin and CompoundHashJoin.

### Important Notes

Creates SimpleHashJoin and CompoundHashJoin.

CompoundHashJoin is what was HashJoin. 
SimpleHashJoin is a new implementation that only indexs the smaller of the 2 tables being joined together.

The rest is refactor and clean-up of the shared join code.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: std-bits/table/src/main/java/org/enso/table/data/table/join/JoinStrategy.java [hunk 1]
```diff
@@ -33,9 +34,9 @@ static JoinStrategy createStrategy(List<JoinCondition> conditions, JoinKind join
       assert !betweenConditions.isEmpty();
       return new SortJoin(betweenConditions, builderSettings);
     } else if (betweenConditions.isEmpty()) {
-      return new HashJoin(hashableConditions, new MatchAllStrategy(), builderSettings);
+      return new SimpleHashJoin(hashableConditions, new MatchAllStrategy(), builderSettings);
     } else {
-      return new HashJoin(
+      return new CompoundHashJoin(
           hashableConditions, new SortJoin(betweenConditions, builderSettings), builderSettings);
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/table/join/JoinStrategy.java]: I feel like a have a de-ja-vu - I think this is similar to what this was before one of my refactors 😅 when I was trying to limit duplicated logic.

But I think that with additional optimizations it indeed makes sense to split back.

File: std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/SimpleHashJoin.java [hunk 2]
```diff
@@ -20,8 +20,8 @@
  * <p>It then delegates to {@code remainingMatcher} to perform the remaining conditions on the
  * matching pairs of row subsets.
  */
-public class HashJoin implements JoinStrategy {
-  public HashJoin(
+public class SimpleHashJoin implements JoinStrategy {
+  public SimpleHashJoin(
       List<HashableCondition> conditions,
       PluggableJoinStrategy remainingMatcher,
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/SimpleHashJoin.java]: Why does this class still contain `remainingMatcher`??

From what I understood the `SimpleHashJoin` would be used for cases where there are no 'remaining matchers'.

Right now, I don't see _any_ difference between `SimpleHashJoin` and `CompoundHashJoin` apart from the names - it seems as if the code was exactly the same? That does not seem right.

Shouldn't `SimpleHashJoin` be simplified as part of this refactor? I know that this is just preparation for further modifications, but I don't think we should be entering a state where we have 2 classes which are essentially copies of one another.

**Review by radeusgd — COMMENTED** [review]
It feels like the `SimpleHashJoin` should be simplified a bit further to justify the splitting.

~~Currently it looks like just splitting two classes and duplicating all of their code which does not seem right.~~

_Edit:_ I see that this was the intent. I somehow missed the description 😅 Still I think we should at least remove the `remainingMatcher` from `SimpleHashJoin`, to make it distinct from `Compound` one. It's a relatively simple change still.

**Review by GregoryTravis — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/CompoundHashJoin.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.enso.table.data.table.join.between.SortJoin;
+import org.enso.table.data.table.join.conditions.Between;
+import org.enso.table.data.table.join.conditions.HashableCondition;
+import org.enso.table.problems.ProblemAggregator;
+import org.graalvm.polyglot.Context;
+
+/**
+ * A strategy that uses a hash-map to perform join on the equality conditions.
+ *
+ * <p>It then delegates to {@code sortJoin} to perform the remaining conditions on the matching
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/CompoundHashJoin.java]:
```suggestion
 * <p>It then delegates to {@code SortJoin} to perform the remaining conditions on the matching
```

nit: shouldn't this be the class name so uppercase?

File: std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/SimpleHashJoin.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  private final HashJoinConfig hashJoinConfig;
+  private final JoinKind joinKind;
+  private final SimpleHashJoinResultBuilder resultBuilder;
+
+  @Override
+  public JoinResult join(ProblemAggregator problemAggregator) {
+    // algorithm assumes that left table is the big table.
+    // If not we have flipped the tables round to do the join.
+    // If you are debugging your left table might not be your left table here.
+    // The result builder flips the indexes back as you add them
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/SimpleHashJoin.java:52]: I appreciate this comment so much ❤️❤️❤️

**Review by radeusgd — APPROVED** [review]

File: std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/CompoundHashJoin.java (hunk 3 above)
Comment (AdRiley) [reply std-bits/table/src/main/java/org/enso/table/data/table/join/hashing/CompoundHashJoin.java]: Yes. Fixed :)

**Comment by radeusgd** [comment]
@AdRiley did you have a chance to run the benchmarks and see how the new algorithm impacts performance? Would be great to see how much faster it is than the older version.

**Comment by AdRiley** [comment]
> @AdRiley did you have a chance to run the benchmarks and see how the new algorithm impacts performance? Would be great to see how much faster it is than the older version.

Only locally on my machine, which I know has a lot of variance. So really I want to see the official figures after merge. But the rough figures from my machine for those 2 new benchmarks we added for this optimisation - The benchmarks run in about 20% of the time of what they did prior to this change. Which is a nice speed up :)

**Comment by radeusgd** [comment]
> So really I want to see the official figures after merge.

Ok, although just FYI it is possible to schedule a CI benchmark run on an unmerged PR.

You can go here: https://github.com/enso-org/enso/actions/workflows/std-libs-benchmark.yml
And manually trigger the run, selecting your branch:
image

> But the rough figures from my machine for those 2 new benchmarks we added for this optimisation - The benchmarks run in about 20% of the time of what they did prior to this change. Which is a nice speed up :)

But that seems like a good enough check to be getting this merged :) I doubt the CI figures would differ much.

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · Workflow runs · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by AdRiley** [comment]
> > So really I want to see the official figures after merge.
> 
> Ok, although just FYI it is possible to schedule a CI benchmark run on an unmerged PR.
> 
> You can go here: https://github.com/enso-org/enso/actions/workflows/std-libs-benchmark.yml And manually trigger the run, selecting your branch: image
> 
> > But the rough figures from my machine for those 2 new benchmarks we added for this optimisation - The benchmarks run in about 20% of the time of what they did prior to this change. Which is a nice speed up :)
> 
> But that seems like a good enough check to be getting this merged :) I doubt the CI figures would differ much.
> 
> >  GitHub**[Benchmark Standard Libraries · Workflow runs · enso-org/enso](https://github.com/enso-org/enso)**Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.

Ah cool! Thank you :) I learnt something new!

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · Workflow runs · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Pull request merged**